File: .scripts/transform.py
from pathlib import Path
from datetime import datetime

def transform(path: Path):
    last_write_date = path.stat().st_mtime
    last_write_date = datetime.fromtimestamp(last_write_date).strftime("%Y-%m-%d")
    print(f"Transforming {path} last modified at {last_write_date}")
    content = path.read_text()
    title = content.splitlines()[0][1:].strip()
    print(f"Title: {title}")
    content = f"""---
title: "{title}"
pubDate: "{last_write_date}"
---
{content}
"""
    path.write_text(content)
    pass


blogs = Path("./src/content/blog")
for blog in blogs.iterdir():
    if blog.is_dir():
        for blog_file in blog.iterdir():
            if blog_file.is_file():
                transform(blog_file)

File: .scripts/transform.js
import { statSync, readFileSync, writeFileSync, readdirSync, rmSync } from "fs";
import { join } from "path";

function transform(filePath) {
  const stats = statSync(filePath);
  const lastWriteDate = new Date(stats.mtime).toISOString().split("T")[0];
  console.log(`Transforming ${filePath} last modified at ${lastWriteDate}`);

  let content = readFileSync(filePath, "utf-8");
  const title = content.split("\n")[0].substring(1).trim();
  if (title === "--") return;
  console.log(`Title: ${title}`);

  content = `---
title: "${title}"
pubDate: "${lastWriteDate}"
---
${content}`;

  writeFileSync(filePath, content);
}

const blogsDir = "./src/content/blog";
readdirSync(blogsDir).forEach((blog) => {
  const blogPath = join(blogsDir, blog);
  if (statSync(blogPath).isDirectory()) {
    readdirSync(blogPath).forEach((blogFile) => {
      const blogFilePath = join(blogPath, blogFile);
      if (statSync(blogFilePath).isFile()) {
        if (blogFilePath.toLowerCase().includes("readme")) {
          rmSync(blogFilePath);
        }
        transform(blogFilePath);
      }
    });
  }
});
